src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/PolicyFunctions.kt

```diff
@@ -53,14 +57,26 @@ fun <T : ILicenceCondition> getPolicyPlaceholders(conditions: List<T>): HashMap<
 fun <T : ILicenceCondition> licencePolicyChanges(
   conditions: List<AdditionalCondition>,
   policyConditions: List<T>,
-  policyPlaceholders: HashMap<String, HashMap<String, String>>
-): List<LicenceConditionChanges> = conditions.mapNotNull { c ->
-  policyConditions.find { it.code == c.code }?.let { pc ->
-    val textHasChanged = pc.text != c.text
-    val dataChanges = c.data.filter { d -> policyPlaceholders[c.code]?.containsValue(d.field) == false }
-    if (textHasChanged || dataChanges.isNotEmpty()) LicenceConditionChanges(
-      c.code!!, c.sequence, c.text!!, pc.text, dataChanges
-    )
-    else null
+  policyPlaceholders: Map<String, Map<String, String>>
+): List<LicenceConditionChanges> = conditions.mapNotNull { condition ->
+  val match = policyConditions.find { it.code == condition.code }
+  if (match != null) {
+    val textHasChanged = match.text != condition.text
+    val dataChanges = condition.data.filter { d -> policyPlaceholders[condition.code]?.containsValue(d.field) == false }
+    val dataHasChanged = dataChanges.isNotEmpty()
+    when {
+      textHasChanged -> TEXT_CHANGE.of(condition, match.text, dataChanges)
+      dataHasChanged -> NEW_OPTIONS.of(condition, match.text, dataChanges)
+      else -> null
+    }
+  } else {
+    DELETED.of(condition)
```

Comment:
still need to distinguish between REPLACED and DELETED here
probably help to sort based on sequence too.